Mathlib/Tactic/ProdAssoc.lean

```diff
@@ -0,0 +1,150 @@
+def mkProdTree (e : Expr) : MetaM ProdTree :=
+  match e with
+    | .app (.app (.const ``Prod [u,v]) X) Y => do
+        return .prod (← X.mkProdTree) (← Y.mkProdTree) u v
+    | X => do
+      let some u := (← inferType X).type? | throwError "Not a type{indentExpr X}"
+      return .type X u
+
+/-- Given `P : ProdTree` representing an iterated product and `e : Expr` which
+should correspond to a term of the iterated product, this will return
+a list, whose items correspond to the leaves of `P` (i.e. the types appearing in the product),
+where each item is the appropriate composition of `Prod.fst` and `Prod.snd` applied to `e`
+resulting in an element of the type corresponding to the leaf.
+
+For example, if `P` corresponds to `(X × Y) × Z` and `t : (X × Y) × Z`, then this
+should return `[t.fst.fst, t.fst.snd, t.snd]`.
+-/
+def ProdTree.unpack (t : Expr) : ProdTree → MetaM (List Expr)
+  | type _ _ => return [t]
+  | prod fst snd u v => do
+      let fst' ← fst.unpack <| mkAppN (.const ``Prod.fst [u,v]) #[fst.getType, snd.getType, t]
+      let snd' ← snd.unpack <| mkAppN (.const ``Prod.snd [u,v]) #[fst.getType, snd.getType, t]
+      return fst' ++ snd'
+
+/-- This function should act as the "reverse" of `ProdTree.unpack`, constructing
+a term of the iterated product out of a list of terms of the types appearing in the product. -/
+def ProdTree.pack (ts : List Expr) : ProdTree → MetaM Expr
+  | type _ _ => do
+    match ts with
+      | [] => throwError "Can't pack the empty list."
+      | [a] => return a
+      | _ => throwError "Failed due to size mismatch."
+  | prod fst snd u v => do
+    let fstSize := fst.size
+    let sndSize := snd.size
+    unless ts.length == fstSize + sndSize do throwError "Failed due to size mismatch."
+    let tsfst := ts.toArray[:fstSize] |>.toArray.toList
+    let tssnd := ts.toArray[fstSize:] |>.toArray.toList
+    let mk : Expr := mkAppN (.const ``Prod.mk [u,v]) #[fst.getType, snd.getType]
+    return .app (.app mk (← fst.pack tsfst)) (← snd.pack tssnd)
+
+/-- Converts a term `e` in an iterated product `P1` into a term of an iterated product `P2`.
+Here `e` is an `Expr` representing the term, and the iterated products are represented
+by terms of `ProdTree`. -/
+def ProdTree.convertTo (P1 P2 : ProdTree) (e : Expr) : MetaM Expr :=
+  return ← P2.pack <| ← P1.unpack e
+
+/-- Given two expressions corresponding to iterated products of the same types, associated in
+possibly different ways, this constructs the "obvious" function from one to the other. -/
+def mkProdFun (a b : Expr) : MetaM Expr := do
+  let pa ← a.mkProdTree
+  let pb ← b.mkProdTree
+  unless pa.components.length == pb.components.length do
+    throwError "The number of components of {a} and {b} must match."
+  for (x,y) in pa.components.zip pb.components do
+    unless ← isDefEq x y do
+      throwError "Components {x} is not defeq to component {y}."
+  return .lam `t a (← pa.convertTo pb <| .bvar 0) .default
```

Comment:
Is constructing this `.lam` necessary? It looks like `pa.convertTo pb` is all you need to return, unless the point of this is to cache the type of the domain for some reason?

Reply:
I'm not sure what you mean. This is meant to construct the actual function, as an `Expr`, between the products. `pa.convertTo pb` will return a function `Expr -> Expr`, so constructing some `.lam` term will have to be done later, presumably as part of `mkProdEquiv`. I can imagine it being useful to have some way to construct the `Expr` of the function, which this accomplishes.

Reply:
Oh, oops, I got lost in the levels of representation here. I mistook the application for a `.app`.

Passing `.bvar 0` like this isn't exactly safe, unless you are very careful that you never use `.lam` ever again inside the term. You don't, so it's ok I think, but the way you're supposed to do this is
```
withLocalDeclD `t a .default fun fvar =>
  mkLambdaFVars #[fvar] (← pa.convertTo pb fvar)
```
(untested, I might have gotten function names and argument orders wrong). If you don't switch to using `mkLambdaFVars`, please document the invariants why `.bvar 0` is OK.

Reply:
Ah I see. Okay I'll make this change soon. I actually had to fight this `bvar` while I was writing this code, so I know for a fact that this is brittle. Thanks!

Reply:
Ok. I've fixed this now, essentially using your suggestion with `withLocalDeclD`.